packages/framework/ai-collab/src/explicit-strategy/agentEditTypes.ts

```diff
@@ -0,0 +1,175 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type { JsonPrimitive } from "./jsonTypes.js";
+
+/**
+ * TODO: The current scheme does not allow manipulation of arrays of primitive values because you cannot refer to them.
+ * We could accomplish this via a path (probably JSON Pointer or JSONPath) from a possibly-null objectId, or wrap arrays in an identified object.
+ *
+ * TODO: only 100 object fields total are allowed by OpenAI right now, so larger schemas will fail faster if we have a bunch of schema types generated for type-specific edits.
+ *
+ * TODO: experiment using https://github.com/outlines-dev/outlines (and maybe a llama model) to avoid many of the annoyances of OpenAI's JSON Schema subset.
+ *
+ * TODO: without field count limits, we could generate a schema for valid paths from the root object to any field, but it's not clear how useful that would be.
+ *
+ * TODO: We don't supported nested arrays yet.
+ *
+ * TODO: Could omit edit contents for setRoot edits as the tree state is the result (or the other way around).
+ *
+ * TODO: Add a prompt suggestion API!
+ *
+ * TODO: Could encourage the model to output more technical explanations of the edits (e.g. "insert a new Foo after "Foo2").
+ *
+ * TODO: Get explanation strings from o1.
+ *
+ * TODO: Tests of range edits.
+ *
+ * TODO: SetRoot might be obscure enough to make the LLM avoid it. Maybe a general replace edit would be better.
+ *
+ * TODO: Handle 429 rate limit error in streamFromLlm.
+ *
+ * TODO: Add an app-specific guidance string.
+ *
+ * TODO: Give the model a final chance to evaluate the result.
+ *
+ * TODO: Separate system prompt into [system, user, system] for security.
+ *
+ * TODO: Top level arrays are not supported with current DSL.
+ *
+ * TODO: Structured Output fails when multiple schema types have the same first field name (e.g. id: sf.identifier on multiple types).
+ *
+ * TODO: Pass descriptions from schema metadata to the generated TS types that we put in the prompt
+ */
+
+/**
+ * TBD
```

Comment:
@taylorsw04 @noencke Suggestions for this?

Reply:
I want to make sure these get documented before we merge into main, otherwise they'll probably be lost 🫤

Reply:
I've given a go at adding jsdoc for these types, please still review and confirm the jsdoc. There are also some remarks asking what different thing are.

Reply:
This is the field we force the LLM to generate to avoid any type ambiguity (e.g. a vector and a point both have x/y and are ambiguous without the LLM telling us which it means).

Reply:
Added JSDOC for this in commit b5cceaa3be6163e6a448c08a5b34afbae6af4ee9